blocks/basic/include/gnuradio-4.0/basic/FunctionGenerator.hpp

```diff
@@ -0,0 +1,297 @@
+clockSrc.addTimeTagEntry(30'000, "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=3");
+@endcode
+
+Secondly, create table to match command and functions:
+@code
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=1" } }, createConstPropertyMap(5.f));
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=2" } }, createLinearRampPropertyMap(5.f, 30.f, .2f));
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=3" } }, createConstPropertyMap(30.f));
+@endcode
+
+Lastly create a matching predicate with the following signature:
+@code
+std::function<std::optional<bool>(const property_map &, const property_map &, std::size_t)>;
+@endcode
+
+
+2) Using user defined ready-to-use Tags.
+@code
+auto &clockSrc = testGraph.emplaceBlock<gr::basic::ClockSource<float>>({ { "sample_rate", 100.f }, { "n_samples_max", 100 }, { "name", "ClockSource" } });
+clockSrc.tags = { Tag(0, createConstPropertyMap(5.f)),
+                  Tag(100, createLinearRampPropertyMap(5.f, 30.f, .2f)),
+                  Tag(300, createConstPropertyMap(30.f)),
+                  Tag(350, createParabolicRampPropertyMap(30.f, 20.f, .1f, 0.02f)),
+                  Tag(550, createConstPropertyMap(20.f)),
+                  Tag(650, createCubicSplinePropertyMap(20.f, 10.f, .1f)),
+                  Tag(800, createConstPropertyMap(10.f)),
+                  Tag(850, createImpulseResponsePropertyMap(10.f, 20.f, .02f, .06f)) };
+@endcode
+
+)"">;
+
+template<typename T>
+    requires(std::floating_point<T>)
+struct FunctionGenerator : public gr::Block<FunctionGenerator<T>, BlockingIO<true>, FunctionGeneratorDoc> {
+    /**
+     * TODO: Taken from CtxSettings
+     * A predicate for matching two contexts
+     * The third "attempt" parameter indicates the current round of matching being done.
+     * This is useful for hierarchical matching schemes,
+     * e.g. in the first round the predicate could look for almost exact matches only,
+     * then in a a second round (attempt=1) it could be more forgiving, given that there are no exact matches available.
+     *
+     * The predicate will be called until it returns "true" (a match is found), or until it returns std::nullopt,
+     * which indicates that no matches were found and there is no chance of matching anything in a further round.
+     */
+    using MatchPredicate = std::function<std::optional<bool>(const property_map &, const property_map &, std::size_t)>;
+
+    PortIn<T>  in; // ClockSource input
+    PortOut<T> out;
+
+    Annotated<float, "sample_rate", Visible, Doc<"sample rate">>                        sample_rate = 1000.f;
+    Annotated<std::string, "signal_type", Visible, Doc<"see function_generator::Type">> signal_type = "Const";
+
+    // Parameters for different functions
+    Annotated<T, "start_value", Visible>                                               start_value    = T(0.);
```

Comment:
Similar to the `SampleRate` above but these should be derived from an enum and using statement that is declared in this function generator namespace.

Reply:
N.B. review whether each of this settings needs to be 'Visible`. The idea behind the `Visible` is that it's and UI hint whether these values should be drawn on the top-level block representation. This is regardless of the accessibility/visibility of these fields within the contextt/block settings UI widgets. I.e. 'Visible` should be parameter that the user must see under any other conditions.

I'd say that there -- ideally -- should be as few as possible, mostly only 4-6 parameters total and more only for justified cases (e.g. digitizer, SDR sources etc.).

Reply:
I removed the Visible flag from all function parameters.